### PR TITLE
daemon: snapd API events endpoint redux

### PR DESCRIPTION
**Pull request by stevenwilkin**
Reworking of #530 but with random strings of alphanumeric characters for subscriber identifiers

### CODE REVIEW

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by niemeyer** [comment]
This looks reasonable. I haven't looked enough into this events systems yet, though. Hopefully someone else has been following through, as we already have pieces of it merged. Might be good to have a higher-level document explaining the design before we go too far.

LGTM, though.

**Comment by mvo5** [comment]
Code looks good to me, for the design and how it fits I would prefer if @chipaca could comment as he has most experience with the daemon.

File: daemon/response.go [hunk 1]
```diff
@@ -27,8 +27,11 @@ import (
 	"path/filepath"
 	"strconv"
 
+	"github.com/gorilla/websocket"
```
Comment (zyga) [inline daemon/response.go]: Do we need to have golang-github-gorilla-websocket in debian/control somewhere or is golang-websocket-dev enough?

File: docs/rest.md [hunk 2]
```diff
@@ -661,3 +661,25 @@ Sample input:
     “slot”: {“snap”: “keyboard-lights”, “slot”: “capslock-led”}
 }
 ```
+
+## /2.0/events
+
+### GET
+
+* Description: Websocket upgrade
```
Comment (zyga) [inline docs/rest.md]: The description is a bit vague IMHO. Websocket upgrade is the mechanism that happens there but surely that's not what this method is doing.

**Comment by zyga** [comment]
+1 with two questions that can be addressed later.

File: daemon/response.go (hunk 1 above)
Comment (chipaca) [inline daemon/response.go]: Package is `golang-websocket-dev`, and yes we need it added to `debian/control`. Also `dependencies.tsv` needs updating, ideally pointing at the same one we'd get from `golang-websocket-dev/xenial`.

File: notifications/hub.go [hunk 3]
```diff
@@ -36,6 +36,11 @@ func NewHub() *Hub {
 	}
 }
 
+// SubscriberCount returns the number of subscribers
+func (h *Hub) SubscriberCount() int {
+	return len(h.subscribers)
```
Comment (chipaca) [inline notifications/hub.go]: you need go grab the lock here, as it's a public method that looks at `h.subscribers`

**Comment by chipaca** [comment]
Can't wait to see this hooked up to Tasks :-)
Looks good in general, the lock thing is a blocker but otherwise good to go.

File: notifications/hub.go (hunk 3 above)
Comment (stevenwilkin) [inline notifications/hub.go]: I thought I could get away without it but you're right, it's part of the public API. Have added it in.

File: docs/rest.md (hunk 2 above)
Comment (stevenwilkin) [inline docs/rest.md]: That was lifted from the [Snappy over REST](https://docs.google.com/document/d/1VqrEK3V_7jYMICy09tOVSaATb5KWu6Vs3CbDu9Nh1cQ/edit#heading=h.d10cqykq8uh0) doc. Would the following be better or is it too much?

"upgrade the connection to a websocket on which notifications will be sent"

**Comment by stevenwilkin** [comment]
Feedback from @chipaca addressed.

**Comment by mvo5** [comment]
This looks good, but the `debian/control` and `dependencies.tsv` both need an update for the new websocket package, so please only merge once that landed or it will break our debian builds.

**Comment by stevenwilkin** [comment]
This is in `dependencies.tsv`:
```
github.com/gorilla/websocket    git 234959944d9cf05229b02e8b386e5cffe1e4e04a    2016-01-28T16:48:56Z
```
This is in `debian/control`:
```
golang-websocket-dev
```
Is it a specific revision I'm missing here or something else @mvo5 ?

**Comment by mvo5** [comment]
@stevenwilkin Ups, sorry! I missed that its already in master. So all good from my side.

**Pull request merged**